compiler/passes/scopeResolve.cpp

```diff
@@ -1961,140 +1976,285 @@ bool lookupThisScopeAndUses(const char*           name,
-
-          for_actuals(expr, block->useList) {
-            // Ensure we only have use or import statements in this list
-            if (UseStmt* use = toUseStmt(expr)) {
-              moduleUses->add(use);
-            } else if (ImportStmt* import = toImportStmt(expr)) {
-              moduleUses->add(import);
-            } else {
-              INT_FATAL("Bad contents for useList, expected use or import");
-            }
-          }
+    // check private use only, not including names of modules
+    // (this forms the 1st shadow scope for a private use)
+    visitedModules.clear();
+    lookupUseImport(name, context, scope, symbols,
+                    renameLocs, storeRenames, reexportPts, visitedModules,
+                    /* forShadowScope */ true,
+                    /* publicOnly */ false);
+  }
 
-          INT_ASSERT(moduleUses->n);
+  if (symbols.size() == 0) {
+    // Check to see if the module name matches what is use'd
+    // with a private use
+    // (this forms the 2nd shadow scope for a private use)
+    lookupUsedImportedMod(name, context, scope, symbols,
+                          renameLocs, storeRenames,
+                          /* forShadowScope */ true,
+                          /* publicOnly */ false);
+  }
 
-          buildBreadthFirstModuleList(moduleUses);
+  return symbols.size() != 0;
+}
 
-          if (enableModuleUsesCache)
-            moduleUsesCache[block] = moduleUses;
+static void lookupUseImport(const char*           name,
+                            BaseAST*              context,
+                            BaseAST*              scope,
+                            std::vector<Symbol*>& symbols,
+                            std::map<Symbol*, astlocT*>& renameLocs,
+                            bool storeRenames,
+                            std::map<Symbol*, VisibilityStmt*>& reexportPts,
+                            VisitedModulesSet& visitedModules,
+                            bool forShadowScope,
+                            bool publicOnly) {
+  // Nothing found so far, look into the uses.
+  if (BlockStmt* block = toBlockStmt(scope)) {
+    if (block->useList != NULL) {
+
+      // check to see if we have already visited this scope
+      // if it is a module scope (to avoid infinite loop with recursive use)
+      if (ModuleSymbol* mod = toModuleSymbol(block->parentSymbol)) {
+        if (mod->block == block) {
+          // if the module is private, return,
+          // since we can't access anything in it
+          // TODO: does this need to have a more involved check?
+          // could a private outer module use a private inner
+          // module and see it?
+          if (publicOnly && mod->hasFlag(FLAG_PRIVATE)) {
+            return;
+          }
+
+          auto pair = visitedModules.insert(std::make_pair(mod, name));
+          if (pair.second == false) {
+            // module+name has already been visited by this function
+            // so don't try to visit it again
```

Comment:
Thanks. Since testing seems to be going OK with this PR I am inclined to agree with you (that it worked ok). However we can watch out for errors along these lines in the future.